### ISSUE TITLE
Page extension overriding field ToolTip doesn't work if language is English (United Kingdom)

### ISSUE THREAD

**Issue by dtkb** [microsoft/AL#6609]
similar to https://github.com/microsoft/AL/issues/5429

```al
pageextension 50100 CustomerListExt extends "Customer List"
{
    layout
    {
        modify(Name)
        {
            ToolTip = 'Can you see me...';
        }
    }
}
```

with English (United Kingdom) - keeps the original ToolTip:
![image](https://user-images.githubusercontent.com/63455896/116722047-dadd5a80-a9d5-11eb-8e48-9f5cc098e29a.png)

with English (United States) or e.g. Dutch (Netherlands) - takes the extended one:
![image](https://user-images.githubusercontent.com/63455896/116721979-d022c580-a9d5-11eb-8fcc-a206f8ac5dc8.png)

AL extension:
```
Name: AL Language
Id: ms-dynamics-smb.al
Description: AL development tools for Dynamics 365 Business Central
Version: 7.1.453917
Publisher: Microsoft
VS Marketplace Link: https://marketplace.visualstudio.com/items?itemName=ms-dynamics-smb.al
```

running against BC 17.5 MS cloud sandbox.

**Comment by atoader** [comment]
The language used for captions is en-us. If you want to use another language, you need to use an XLIFF file with translations.

**Comment by dtkb** [comment]
I don't think you understood the thread... or I don't think I understand how any of this is meant to work.

The issue is that overriding a tooltip with the **non-ML** `ToolTip` property does not work in en-GB and possibly in other languages. Surely they should default to the untranslated text from the extension property?

Are you really saying that for new captions/tooltips to work in languages other than en-US, we need to create an XLIFF file and translate them? Shouldn't those other languages fallback to the en-US / non-ML text if no specific translation of their own is available?

**Comment by dtkb** [comment]
and if en-GB isn't *meant* to work like this, then why did Dutch (Netherlands) ??

Please re-read and open if relevant, or if this really isn't meant to work, consistency and documentation is direly needed.

**Comment by atoader** [comment]
en-US is the default language for captions specified as "ToolTip". The runtime should default to en-US if a caption for the current language is not found, but that is a runtime issue, and not a developer tools issue. As stated in the readme, those need to be addressed through support.

**Comment by dtkb** [comment]
OK, I've opened a bug with our support partner, and hopefully they can get it to the relevant team.

It would be great if there could be a 'fast track' for that, e.g. if the platform team (or whoever) had a private GitHub repo, to which you could transfer issues that are verifiable. I understand they don't want an open repo for every random BC/NAV/Navision/Attain user in the world to ask how to add posting groups... but having to go through partners is tough and sometimes futile, at least if your partner says 'it has been like that forever, so I guess MS will not care, closing'... sigh.

**Comment by atoader** [comment]
I totally understand. As far as I know, there aren't any plans to expand GitHub to other teams so this is the only way to get this issues fixed.

**Comment by dtkb** [comment]
Our partner advise that "a ticket has been raised with Microsoft Technical support under ticket no: 2105060050001848" 🥳 🤞

**Comment by dtkb** [comment]
updates from my ticket with MS:

MS:
> Just wanted to give you an update, during my tests I was able to replicate the same behavior as you mentioned.
>
> I tested in tenants with different Localization (UK and BE) and when changing the Region and Language for not being English US but for the specific environment localization language, the behavior is like you described.
> 
> Additionally I was analyzing if the environment would have any language app installed that could cause this, but everything is clear in regards to that.
>
> I will share these findings with our team responsible for this development, it seems that this bug reappeared again.
> 
> Once I have more details from our development team I will let you know._

me:
> Just to check, did you test/replicate both the PageExtension->ToolTip and TableExtension->Caption problems? As I guess they might have a similar root cause, so maybe best to report/fix both at the same time. If you didn’t already of course :-D

MS:
> Thank you for the heads up in regards to TableExtension->Caption. I just tested right now and I’m experience the same bug as you.
>
> I have just addressed the same to our team to have this into consideration as well.
>
> Both these 2 scenarios (PageExtension->ToolTip and TableExtension->Caption) were tested and addressed. And you are right, report both at the same time will be better for all. The same team will be handling these 2 scenarios.
>
> At this moment I’m waiting for their feedback, once I have any details I will let you know.

**Comment by dtkb** [comment]
I was invited to a meeting with MS where they shared a PowerPoint that explained the current order of resolution/precedence of captions/translations between various sources. I don't fully understand it yet, but I **think** it does indicate this is intended behaviour, at least in my cases.

Basically,
* when providing non-ML Caption/ToolTip/etc., that's treated as en-US. And en-US (or, failing that, object/field name) is the ultimate fallback if no 'better' translations are available.
* ...but if I now deploy that in an extension, to a tenant running the en-GB localised base app, and view it as a user whose Language is en-GB...
* ...the rules of precedence dictate that the en-GB translation from the base app wins, because I'm a user set to show en-GB - so suppressing my en-US one from the extension.

MS are going to document this properly (and perhaps tweak the logic later, if demand arises) so that at least no one else will get confused. Watch this space! ;-)

**Comment by dtkb** [comment]
Here is MS's summary, followed by screengrabs of their PowerPoint presentation, with tables of the 'weighting'/precedence of different sources of translations, with numerically higher ones winning out.

Hope this helps others too!

***

We have the possibility to share with you a little test scenario with Caption and Tooltip and also explaining how the translation works.

Attached you will find the slide presentation we used in our teams meeting.

We explained both scenarios highlighted on this ticket.

Let’s share our explanation for the ToolTip.

When you are using, what we call, a label property like Caption or Tooltip, then the language it is going to target is English-United States (ENU), which means that

    Caption = ‘My English Text’;

Is the same as

    CaptionML = ENU = ‘My English Text’;

You have a GB tenant which comes with English-Great Britain (ENG) translations.

So when the system has to pick a translation it tries to find the best match, and in the mix of possible translations it finds

* ENU = ‘Test’
* ENG = ‘Specifies the ….’
* And maybe more.

So when you set the language in settings to ENG, then it starts out by looking at can it find a translation with ENG, and it does so the text ‘Specifies the…’ is shown.

When you then change to ENU then it also get’s a match on the user specified language and text ‘Test’ is shown..

Now the explanation for the Caption.

When it comes to finding the translation for Page Controls then it uses the same logic as for Page Actions, with a small twist. If it can’t find any translation that matches the language the user has set, including the fallback to the primary language of the user specified one, and ENU (ENU is hard coded) then it looks at the table field that is referenced in the page control.

The control called ‘Name’ doesn’t have a caption on it, but all table fields does, so the table field has a translation for ENG.
When you then write a caption on the page control, then it is getting caught by the fallback logic, before it starts to look at the table fields.

And last for the Caption on the TableExtensions modify field, is the same story as for the ToolTip.

All in all, the system works as intended. It might not be the most intuitive system as we shared, but we are working with docs writer about making a page about how all of this works, so all our Customers and Partners better understand it.

***

![image](https://user-images.githubusercontent.com/63455896/118833313-428e1380-b8b9-11eb-90cf-1912d3b7da1a.png)

![image](https://user-images.githubusercontent.com/63455896/118833380-4cb01200-b8b9-11eb-9821-b6747d068c08.png)

![image](https://user-images.githubusercontent.com/63455896/118833414-546fb680-b8b9-11eb-8487-cd1c736ec3c9.png)
